[AArch64] Run optimizeTerminators earlier too.

davemgreen opened the pull request:
Running optimizeTerminators prior to other optimizations like branch layout can lead to more folding and better codegen, but is not on its own able to capture all cases. There is benefit to running it in both places. This adds the existing code from #161508 into the AArch64RedundantCopyElimination pass, which sounds like a sensible enough place for it.

hassnaaHamdi commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -11587,6 +11587,49 @@ AArch64InstrInfo::analyzeLoopForPipelining(MachineBasicBlock *LoopBB) const {
       Init, IsUpdatePriorComp, Cond);
 }
 
+bool llvm::optimizeTerminators(MachineBasicBlock *MBB,
```
Could you move the function to the end of the file? because it's between 2 class(AArch64InstrInfo) member functions while it's not related.

hassnaaHamdi reviewed (commented):
What about cases like this:
```
mov w10, wzr
mul x8, x0, x2
stp x8, x9, [x4]
cbz w10, .LBB2_2
```
Will it be supported in near future ? I think it’s very similar to the case you are handling, right?

davemgreen reviewed (commented):
> What about cases like this:
> 
> mov w10, wzr
> mul x8, x0, x2
> stp x8, x9, [x4]
> cbz w10, .LBB2_2

Do you have an example where you see that come up? Ideally we would propagate the wzr to the cbz, and then this optimization should kick-in.

davemgreen committed:
[AArch64] Run optimizeTerminators earlier too.

Running optimizeTerminators prior to other optimizations like branch layout can
lead to more folding and better codegen, but is not on its own able to capture
all cases. There is benefit to running it in both places. This adds the
existing code from #161508 into the AArch64RedundantCopyElimination pass, which
sounds like a sensible enough place for it.

usha1830 reviewed (approved):
LGTM, Thanks.

dtellenbach reviewed (approved):
LGTM, thanks!

aeubanks commented:
I've reverted in https://github.com/llvm/llvm-project/pull/171505

davemgreen commented:
I'll take a look and see what might be going wrong - thanks.

googlewalt commented:
Thanks for the quick fix!  I verified that fixed our issue.
